### ISSUE TITLE
Closed{Add,Sub} trait-bound is unnecessarily strict for Point::{add,sub}

### ISSUE THREAD

**Issue by z33ky** [rustsim/nalgebra#719]
As background for this issue, I've implemented two newtypes `Coordinate` and `Distance` to wrap an `f32` and use these as basis for nalgebra's `Point` and `Vector` types:
```
use derive_more::{Add, AddAssign, Display};

#[derive(Clone, Copy, Debug, Display, PartialEq, PartialOrd)]
pub struct Coordinate(pub f32);

#[derive(Add, AddAssign, Clone, Copy, Debug, Display, PartialEq, PartialOrd)]
pub struct Distance(pub f32);

impl std::ops::Mul<f32> for Distance { ... }
impl std::ops::MulAssign<f32> for Distance { ... }

pub use type Point = nalgebra::Point2<Coordinate>;
pub use type Vector = nalgebra::Vector2<Distance>;
```
I have then encountered that subtracting two `Point`s does not work.

I think requiring `N: ClosedSub` for `impl Sub<Point<N, D>> for Point<N, D>` is too strict.
`ClosedSub` requires that `N: Sub<N, Output=N> + SubAssign<N>`. I see no reason why `SubAssign` is required.
Further, the `Output=N` for `Sub` seems unnecessarily restrictive. The `Sub`-impl returns `VectorSum<N, D, D>`, which would require that bound, however if it returned `VectorSum<Sub<N>::Output, D, D> (where Sub<N>::Output: Scalar)` then `Output` would be allowed to differ from `N`.

I haven't gotten through the generics for `Point::add`, but adding my `Vector` to a `Point` also fails, so I'm assuming that `Point::add` has a similar situation and could be resolved similarly.

The same restriction seems to apply to other types too, e.g. `Vector`, though that is currently not an issue for me.

**Comment by sebcrozet** [comment]
Hi!

Yes, nalgebra uses `Closed{Sub, Add, Mul, Div}` almost systematically instead of just `Sub/Add/Mul/Div`. The main reason is to simplify our code base. In particular, allowing an `Output` different than `N` itself would make some impls more verbose and difficult.

Relaxing this restriction whenever it is not too difficult still represent quite a lot of work though so any help is welcome, even if only a few methods are addressed.

**Comment by z33ky** [comment]
I'm starting to remove the need for `{Add,Sub}Assign` for `impl {Add,Sub} for Matrix`.
I have it working for `Point` so far, I'm sure I'll be able to extend it to the other types. I'll open a PR for that once it's complete.
I'll look into `Output=!Self` after that, if I get that working I'll open another PR and I'll have what I need at the moment.

When that is done it seems like enhancing `Mul` and `Div`, as well as allowing `Right=!Self` wouldn't be far off, so I'll likely tackle them as well if I am successful.
One step at a time though :)

**Comment by z33ky** [comment]
I have now also started enhancing the implementation to allow `Output=!Self`.
Arithmetic on the `Matrix` reuses the storage if at least one parameter is passed by value. This is not possible if `Output=!Self` (since a `Matrix<T, ...>` is not a `Matrix<Output, ...>`) such that new storage will always need to be allocated instead. As having `Output=!Self` is probably the uncommon case, this is not a desirable solution.
One option would be only allowing `Output=!Self` when both parameters are references, as storage will need to be allocated whether `Output=Self` or not. In code that means the following (assuming the type definitions in the OP):
```
let a: Point = ...;
let b: Point = ...;
let c = a - b; //error
let c = &a - b; //error
let c = a - &b; //error
let c = &a - &b; //ok
```
All four cases would still work for `nalgebra::Point2<f32>` though. Would that be an acceptable limitation?

I've also started on `Mul` and `Div`. There's also an issue with that, specifically the operand order.
`2.0 * vector` can work, but `vector * 2.0` will not when preserving existing `impl`s.
Currently, there's an `impl<N, ...> Mul<N> for Matrix<N, ...> where N: ClosedMul`, which would allow e.g. `vector * Distance(2.0)` (if one implements `ClosedMul` for `Distance`).
To make `vector * 2.0` work, I would need the rhs-equivalence to `left_scalar_mul_impl!`, which of course would collide with the existing `impl Mul for Matrix`.
Do you think removing that `impl` may be acceptable, in favor of providing an `impl` only for the primitive number types? This would mean that `vector * Distance(2.0)` wouldn't work anymore, but `vector * 2.0` would.

**Comment by z33ky** [comment]
Hey @sebcrozet,
I just want to make sure this isn't accidentally forgotten.
I can try expanding the explanations of the issues and possible workarounds if something's unclear.

**Comment by sebcrozet** [comment]
Hi @z33ky! Thank you for pinging me on this issue.

> All four cases would still work for nalgebra::Point2<f32> though. Would that be an acceptable limitation?

Yes, that looks acceptable. This is a bit confusing but better than nothing for now. I guess we will be able to make every case work once specialization lands.

> I've also started on Mul and Div. There's also an issue with that, specifically  the operand order.
> 2.0 * vector can work, but vector * 2.0 will not when preserving existing impls.
Currently, there's an impl<N, ...> Mul<N> for Matrix<N, ...> where N: ClosedMul, which would allow e.g. vector * Distance(2.0) (if one implements ClosedMul for Distance).
> To make vector * 2.0 work, I would need the rhs-equivalence to left_scalar_mul_impl!, which of course would collide with the existing impl Mul for Matrix.
> Do you think removing that impl may be acceptable, in favor of providing an impl only for the primitive number types? This would mean that vector * Distance(2.0) wouldn't work anymore, but vector * 2.0 would.

I don't think that removing the blanket impl `impl<N, ...> Mul<N> for Matrix<N, ...> where N: ClosedMul` is an option because that would make it impossible to multiply a matrix by a scalar in any generic code.
Also I think I did not fully understand the problem. Why can't we just replace the existing mul impl by:

```rust
impl <N, ...> Mul<N> for Matrix<N, ...>
where N: Mul<N> {
    type Output = Matrix<<N as Mul<N>>::Output>;
    ...
}
```
?

**Comment by z33ky** [comment]
> I don't think that removing the blanket impl `impl<N, ...> Mul<N> for Matrix<N, ...> where N: ClosedMul` is an option because that would make it impossible to multiply a matrix by a scalar in any generic code.
> Also I think I did not fully understand the problem. Why can't we just replace the existing mul impl by:
> 
> ```rust
> impl <N, ...> Mul<N> for Matrix<N, ...>
> where N: Mul<N> {
>     type Output = Matrix<<N as Mul<N>>::Output>;
>     ...
> }
> ```
> 
> ?

That exists in the current code (in master) and allows `vector * Distance(2.0)` (since`N = Distance`), but I would want to use `vector * 2.0`.

I've tried `impl<N, M, ...> Mul<M> for Matrix<N, ...> where N: Mul<M>` but that creates conflicts with `impl Mul<Matrix<*>> for Matrix<*>`.

Adding an `impl<N, ...> Mul<f32> for Matrix<N, ...>` (essentially what `left_scalar_mul_impl!` does to allow `f32 * Matrix<Distance, ...>`) creates conflicts with the existing `impl<N, ...> Mul<N> for Matrix<N, ...>`.
